master/pkg/stream/stream.go

```diff
@@ -0,0 +1,205 @@
+	var lock sync.Mutex
+	cond := sync.NewCond(&lock)
+	if prepareFn == nil {
+		panic("PrepareFn must be provided")
+	}
+	return &Streamer{Cond: cond, PrepareFn: prepareFn}
+}
+
+// Close closes a streamer.
+func (s *Streamer) Close() {
+	s.Cond.L.Lock()
+	defer s.Cond.L.Unlock()
+	s.Cond.Signal()
+	s.Closed = true
+}
+
+// Subscription manages a streamer's subscription to messages of type T.
+type Subscription[T Msg] struct {
+	// Which streamer is collecting messages from this Subscription?
+	Streamer *Streamer
+	// Which publisher should we connect to when active?
+	Publisher *Publisher[T]
+	// Decide if the streamer wants this message.
+	filter func(T) bool
+	// Decide if the streamer has permission to view this message.
+	permissionFilter func(T) bool
+	// wakeupID prevent duplicate wakeups if multiple events in a single Broadcast are relevant
+	wakeupID int64
+}
+
+// NewSubscription creates a new Subscription to messages of type T.
+func NewSubscription[T Msg](
+	streamer *Streamer,
+	publisher *Publisher[T],
+	permFilter func(T) bool,
+	filterFn func(T) bool,
+) Subscription[T] {
+	return Subscription[T]{
+		Streamer:         streamer,
+		Publisher:        publisher,
+		permissionFilter: permFilter,
+		filter:           filterFn,
+	}
+}
+
+// Register a Subscription with its Publisher.
+func (s *Subscription[T]) Register() {
+	s.Publisher.Lock.Lock()
+	defer s.Publisher.Lock.Unlock()
+	s.Publisher.Subscriptions[s] = struct{}{}
+}
+
+// Unregister removes a Subscription from its Publisher.
+func (s *Subscription[T]) Unregister() {
+	s.Publisher.Lock.Lock()
+	defer s.Publisher.Lock.Unlock()
+	delete(s.Publisher.Subscriptions, s)
+}
+
+// Publisher is responsible for publishing messages of type T
+// to streamers associate with active subscriptions.
+type Publisher[T Msg] struct {
+	Lock          sync.Mutex
+	Subscriptions map[*Subscription[T]]struct{}
+	WakeupID      int64
+}
+
+// NewPublisher creates a new Publisher for message type T.
+func NewPublisher[T Msg]() *Publisher[T] {
+	return &Publisher[T]{
+		Subscriptions: map[*Subscription[T]]struct{}{},
+	}
+}
+
+// CloseAllStreamers closes all streamers associated with this Publisher.
+func (p *Publisher[T]) CloseAllStreamers() {
+	p.Lock.Lock()
+	defer p.Lock.Unlock()
+	seenStreamersSet := make(map[*Streamer]struct{})
+	for sub := range p.Subscriptions {
+		if _, ok := seenStreamersSet[sub.Streamer]; !ok {
+			sub.Streamer.Close()
+			seenStreamersSet[sub.Streamer] = struct{}{}
+		}
+	}
+	p.Subscriptions = nil
+}
+
+// Broadcast receives a list of events, determines if they are
+// applicable to the publisher's subscriptions, and sends
+// appropriate messages to corresponding streamers.
+func (p *Publisher[T]) Broadcast(events []Event[T]) {
+	p.Lock.Lock()
+	defer p.Lock.Unlock()
+
+	// start with a fresh wakeupid
+	p.WakeupID++
+	wakeupID := p.WakeupID
+
+	// check each event against each subscription
+	for sub := range p.Subscriptions {
+		func() {
```

Comment:
note: look into why this entire section is necessary to be anonymized, should this be only around the defer